google/cloud/storage/internal/signed_url_requests.cc

```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/signed_url_requests.h"
+#include <sstream>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+std::string SignUrlRequest::CreateBlob() const {
+  std::ostringstream os;
+
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(
+      expiration_time_.time_since_epoch());
+
+  os << verb() << "\n"
+     << md5_hash_value_ << "\n"
+     << content_type_ << "\n"
+     << seconds.count() << "\n";
+
+  for (auto const& kv : extension_headers_) {
+    os << kv.first << ":" << kv.second << "\n";
+  }
+  os << "/" << bucket_name() << "/" << object_name();
```

Comment:
We should be URL-encoding the object name here, as done in the Python client lib:
https://github.com/googleapis/google-cloud-python/blob/6b327e9eedae1430c4abe2a628faeb1f96bc590a/storage/google/cloud/storage/blob.py#L352

The Python library uses the `quote` method from `urllib`, which is almost the same as `curl_easy_escape`... except that `curl_easy_escape` also URL-encodes forward slashes.  I think the correct behavior is actually to escape the forward slashes anyway, so curl's version should suffice.

Reply:
The documentation:

https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-resources

suggests that forward slashes should be preserved.  Let me test this and get back to you.

Reply:
There are a number of other characters that are valid to use in object names, but could potentially cause problems in URLs (e.g. the space character, question mark, and ampersand come to mind).

Reply:
Doh!  Indeed.  Thanks for your patience, fixed.